src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -515,46 +516,91 @@ public void ClearValidationState(string key)
             }
         }
 
-        private ModelStateEntry GetModelStateForKey(string key)
+        private ModelStateNode GetNode(string key) => GetNode(key, createIfNotExists: false);
+
+        private ModelStateNode GetOrAddNode(string key) => GetNode(key, createIfNotExists: true);
+
+        private ModelStateNode GetNode(string key, bool createIfNotExists)
         {
-            if (key == null)
+            Debug.Assert(key != null);
+            if (key.Length == 0)
             {
-                throw new ArgumentNullException(nameof(key));
+                return _root;
+            }
+
+            var current = _root;
+            var previousIndex = 0;
+            int index;
+            while ((index = key.IndexOfAny(Delimiters, previousIndex)) != -1)
+            {
+                var keyStart = previousIndex == 0 ? previousIndex : previousIndex - 1;
+                var subKey = new StringSegment(key, keyStart, index - keyStart);
```

Comment:
Yeah, that's correct. I'll add a comment explaining what the structure looks like.